pkg/engine/executor/executor.go

```diff
@@ -76,33 +80,127 @@ func (c *Context) executeDataObjScan(ctx context.Context, node *physical.DataObj
 		return errorPipeline(errors.New("no object store bucket configured"))
 	}
 
-	predicates := make([]logs.RowPredicate, 0, len(node.Predicates))
+	obj, err := dataobj.FromBucket(ctx, c.bucket, string(node.Location))
+	if err != nil {
+		return errorPipeline(fmt.Errorf("creating data object: %w", err))
+	}
+
+	var (
+		streamsSection *streams.Section
+		logsSection    *logs.Section
+	)
+
+	for _, sec := range obj.Sections().Filter(streams.CheckSection) {
+		if streamsSection != nil {
+			return errorPipeline(fmt.Errorf("multiple streams sections found in data object %q", node.Location))
+		}
+
+		var err error
+		streamsSection, err = streams.Open(ctx, sec)
+		if err != nil {
+			return errorPipeline(fmt.Errorf("opening streams section %q: %w", sec.Type, err))
+		}
+	}
+	if streamsSection == nil {
+		return errorPipeline(fmt.Errorf("streams section not found in data object %q", node.Location))
+	}
+
+	for i, sec := range obj.Sections().Filter(logs.CheckSection) {
+		if i != node.Section {
+			continue
+		}
+
+		var err error
+		logsSection, err = logs.Open(ctx, sec)
+		if err != nil {
+			return errorPipeline(fmt.Errorf("opening logs section %q: %w", sec.Type, err))
+		}
+	}
+	if logsSection == nil {
+		return errorPipeline(fmt.Errorf("logs section %d not found in data object %q", node.Section, node.Location))
+	}
+
+	predicates := make([]logs.Predicate, 0, len(node.Predicates))
 
 	for _, p := range node.Predicates {
-		conv, err := buildLogsPredicate(p)
+		conv, err := buildLogsPredicate(p, logsSection.Columns())
 		if err != nil {
 			return errorPipeline(err)
 		}
 		predicates = append(predicates, conv)
 	}
 
-	obj, err := dataobj.FromBucket(ctx, c.bucket, string(node.Location))
-	if err != nil {
-		return errorPipeline(fmt.Errorf("creating data object: %w", err))
-	}
-
-	return newDataobjScanPipeline(dataobjScanOptions{
-		Object:      obj,
+	var pipeline Pipeline = newDataobjScanPipeline(dataobjScanOptions{
+		// TODO(rfratto): passing the streams section means that each DataObjScan
+		// will read the entire streams section (for IDs being loaded), which is
+		// going to be quite a bit of wasted effort.
+		//
+		// Longer term, there should be a dedicated plan node which handles joining
+		// streams and log records based on StreamID, which is shared between all
+		// sections in the same object.
+		StreamsSection: streamsSection,
+
+		LogsSection: logsSection,
 		StreamIDs:   node.StreamIDs,
-		Section:     node.Section,
 		Predicates:  predicates,
 		Projections: node.Projections,
 
-		Direction: node.Direction,
-		Limit:     node.Limit,
+		// TODO(rfratto): pass custom allocator
+		Allocator: memory.DefaultAllocator,
+
+		BatchSize: c.batchSize,
+		CacheSize: 16_000_000, // TODO(rfratto): make configurable at engine level
```

Comment:
When executing a plan scanning 32 objects with 16 sections each, then this would result in 8GB alone for the cache.